File: leetcode/graph/1136_Parallel_Courses/codes/bfs.py
from collections import deque, defaultdict
from typing import List

class Solution:
    def minimumSemesters(self, n: int, relations: List[List[int]]) -> int:
        graph = defaultdict(list)
        in_degree = [0] * (n + 1)

        for u, v in relations:
            graph[u].append(v)
            in_degree[v] += 1

        queue = deque()
        for i in range(1, n + 1):
            if in_degree[i] == 0:
                queue.append(i)

        semesters = 0
        taken = 0

        while queue:
            for _ in range(len(queue)):
                node = queue.popleft()
                taken += 1
                for nei in graph[node]:
                    in_degree[nei] -= 1
                    if in_degree[nei] == 0:
                        queue.append(nei)
            semesters += 1

        return semesters if taken == n else -1

File: leetcode/graph/1136_Parallel_Courses/codes/bfs.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int minimumSemesters(int n, int** relations, int relationsSize, int* relationsColSize) {
    int* in_degree = calloc(n + 1, sizeof(int));
    int** graph = malloc((n + 1) * sizeof(int*));
    int* graphSize = calloc(n + 1, sizeof(int));
    int* queue = malloc((n + 1) * sizeof(int));
    int front = 0, back = 0;

    for (int i = 0; i <= n; i++) {
        graph[i] = malloc((relationsSize + 1) * sizeof(int));
    }

    for (int i = 0; i < relationsSize; ++i) {
        int u = relations[i][0], v = relations[i][1];
        graph[u][graphSize[u]++] = v;
        in_degree[v]++;
    }

    for (int i = 1; i <= n; ++i)
        if (in_degree[i] == 0)
            queue[back++] = i;

    int semesters = 0, taken = 0;

    while (front < back) {
        int size = back - front;
        for (int i = 0; i < size; ++i) {
            int node = queue[front++];
            taken++;
            for (int j = 0; j < graphSize[node]; ++j) {
                int nei = graph[node][j];
                if (--in_degree[nei] == 0)
                    queue[back++] = nei;
            }
        }
        semesters++;
    }

    free(in_degree);
    free(graphSize);
    for (int i = 0; i <= n; i++) free(graph[i]);
    free(graph);
    free(queue);

    return taken == n ? semesters : -1;
}
